### PR TITLE
[RLlib] POC: Run PGTrainer w/o the distr. exec API (Trainer's new training_iteration method).

### PR DESCRIPTION
**Pull request by sven1977**
POC: Run PGTrainer w/o the distr. exec API (Trainer's new training_iteration method).
- For PGTrainer only: Uses a new flag: `_disable_execution_plan_api=True`.
- Trainer: If above flag is set, do not use the LocalIterator-based distr. exec logic, but call the new Trainer.training_iteration method on each training iteration.
- Implement termination logic for iterations (min-time-s, min-sampling-timesteps, min-train-timesteps).
- Implement metrics collection after train iterations.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/agents/trainer.py [hunk 1]
```diff
@@ -791,8 +823,20 @@ def env_creator_from_classpath(env_context):
...
                 config=self.config,
                 num_workers=self.config["num_workers"])
-            self.train_exec_impl = self.execution_plan(
-                self.workers, self.config, **self._kwargs_for_execution_plan())
+
+            # Function defining one single training iteration's behavior.
+            if self.config["_disable_distributed_execution_api"]:
+                # Ensure remote workers are initially in sync with the
+                # local worker.
+                self.workers.sync_weights()
```
Comment (gjoliver) [inline rllib/agents/trainer.py:831]: set self.train_exec_impl to None here?

File: rllib/agents/trainer.py [hunk 2]
```diff
@@ -927,16 +982,105 @@ def step(self) -> ResultDict:
                 # Allow logs messages to propagate.
                 time.sleep(0.5)
                 raise e
+
+            # Stopping criteria.
```
Comment (gjoliver) [inline rllib/agents/trainer.py:986]: why introduce stopping criteria with this PR for only _disable_distributed_execution_api case?

File: rllib/agents/trainer.py [hunk 3]
```diff
@@ -927,16 +982,105 @@ def step(self) -> ResultDict:
...
-            raise RuntimeError("Failed to recover from worker crash.")
-
         if hasattr(self, "workers") and isinstance(self.workers, WorkerSet):
+            # Sync filters on workers.
             self._sync_filters_if_needed(self.workers)
 
+            # Collect worker metrics.
+            if self.config["_disable_distributed_execution_api"]:
+                info = result.copy()
+                result["info"] = {LEARNER_INFO: info}
```
Comment (gjoliver) [inline rllib/agents/trainer.py:1020]: hmm, we recursively save another copy of result under info key??

File: rllib/agents/trainer.py [hunk 4]
```diff
@@ -927,16 +982,105 @@ def step(self) -> ResultDict:
...
+
+                # Collect rollout worker metrics.
+                episodes, self._episodes_to_be_collected = collect_episodes(
```
Comment (gjoliver) [inline rllib/agents/trainer.py:1023]: are we running stats on a new set of episodes?
this looks pretty weird that we collect a bunch of episodes after step_attemp() call, then just drop them?
btw, may be cleaner to put all these summarization code in a helper func.

File: rllib/agents/trainer.py [hunk 5]
```diff
@@ -980,20 +1126,28 @@ def auto_duration_fn(unit, num_eval_workers, eval_cfg, num_units_done):
...
             # Kick off evaluation-loop (and parallel train() call,
             # if requested).
             # Parallel eval + training.
             if self.config["evaluation_parallel_to_training"]:
                 with concurrent.futures.ThreadPoolExecutor() as executor:
                     train_future = executor.submit(
-                        lambda: next(self.train_exec_impl))
+                        lambda: (next(self.train_exec_impl) if
+                                 use_exec_api else
+                                 self.training_iteration()))
```
Comment (gjoliver) [inline rllib/agents/trainer.py:1150]: the same if-else is repeated 3 time here. make a local helper func? and you can simply submit it here.

File: rllib/policy/sample_batch.py [hunk 6]
```diff
@@ -367,17 +375,10 @@ def shuffle(self) -> "SampleBatch":
...
-            for i, p in enumerate(path):
-                if i == len(path) - 1:
-                    curr[p] = value[permutation]
-                # Translate into list (tuples are immutable).
-                if isinstance(curr[p], tuple):
-                    curr[p] = list(curr[p])
-                curr = curr[p]
-
-        tree.map_structure_with_path(_permutate_in_place, self)
+        self_as_dict = {k: v for k, v in self.items()}
```
Comment (gjoliver) [inline rllib/policy/sample_batch.py:378]: is this really necessary? self already looks like a dict.

File: rllib/policy/sample_batch.py [hunk 7]
```diff
@@ -367,17 +375,10 @@ def shuffle(self) -> "SampleBatch":
...
+        shuffled = tree.map_structure(lambda v: v[permutation], self_as_dict)
+        for k, v in shuffled.items():
+            self[k] = v
```
Comment (gjoliver) [inline rllib/policy/sample_batch.py:381]: instead of the for loop, can you just do self.update(shuffled)?

File: rllib/evaluation/worker_set.py [hunk 8]
```diff
@@ -139,17 +139,34 @@ def remote_workers(self) -> List[ActorHandle]:
...
-            for e in self.remote_workers():
-                e.set_weights.remote(weights)
+        if self.local_worker() is None and from_worker is None:
+            raise TypeError(
+                "No `local_worker` in WorkerSet, must provide `from_worker` "
+                "arg in `sync_weights()`!")
+
+        # Only sync if we have remote workers or `from_worker` is provided.
+        if self.remote_workers() or from_worker is not None:
+            weights = ray.put((from_worker
```
Comment (gjoliver) [inline rllib/evaluation/worker_set.py:160]: saving from_worker in obj store? I thought we need weights here.

**Review by gjoliver — COMMENTED** [review]
a few comments.
I guess the biggest question I have after reading this is that if we can handle those more complicated async agents properly with this pattern.
for example: Concurrently([store_op, replay_op, update_op])
what is the ray programming pattern that will handle this well in a complicated de-coupled fashion?

File: rllib/agents/trainer.py (hunk 2 above)
Comment (sven1977) [reply rllib/agents/trainer.py:986]: Because the `execution_plan` handles this by itself via the `StandardMetricsReporting` op at the end. One of the reasons users didn't understand what the `min_iter_time_s` and `timesteps_per_iteration` config params meant and were doing under the hood.

File: rllib/policy/sample_batch.py (hunk 6 above)
Comment (sven1977) [reply rllib/policy/sample_batch.py:378]: We cannot change in place as any tuples in the nested structures are not alterable. I had them converted to lists before (in the old mapping fn), but I didn't like this solution as it changes the actual nested struct (and this caused errors in another location). If we want to change the SampleBatch in place, this is still the cleanest and most concise solution, I think.

Bottom line: `tree.map_structure(lamdba ..., self)` will not work (not w/o a much more complex mapping fn).

File: rllib/policy/sample_batch.py (hunk 7 above)
Comment (sven1977) [reply rllib/policy/sample_batch.py:381]: done

File: rllib/evaluation/worker_set.py (hunk 8 above)
Comment (sven1977) [reply rllib/evaluation/worker_set.py:160]: This is correct (and the logic here wasn't touched in this PR):

1) ray.put places all the to-be-synched new weights in the object store and returns a object ref.
2) This object ref is used in the call to `set_weights.remote(...)`.
Note that if we had called `set_weights.remote()` with the actual weights (numpy), ray would have automatically transferred these to the object store first (and would have had to do so for all n remote workers, which is inefficient).

File: rllib/agents/trainer.py (hunk 5 above)
Comment (sven1977) [reply rllib/agents/trainer.py:1150]: Great catch! Done.

File: rllib/agents/trainer.py (hunk 2 above)
Comment (gjoliver) [reply rllib/agents/trainer.py:991]: oh yes.
let's make this a util function in metric_ops.py, and see if we can pretty much reuse it in all the trainers?

File: rllib/policy/sample_batch.py (hunk 6 above)
Comment (gjoliver) [reply rllib/policy/sample_batch.py:378]: interesting 👌 
thanks for the explanation.

File: rllib/evaluation/worker_set.py (hunk 8 above)
Comment (gjoliver) [reply rllib/evaluation/worker_set.py:160]: oh I mis-read this. I thought we are putting from_worker in the obj store.
really we are just doing (from_worker or self.local_worker()) here. man I hate how yapf likes to break up long lines ...
sorry, and thanks.

File: rllib/agents/pg/default_config.py [hunk 9]
```diff
@@ -13,6 +13,11 @@
 
     # Experimental: By default, switch off preprocessors for PG.
     "_disable_preprocessor_api": True,
+
+    # PG is the first algo (experimental) to not use the distr. exec API
+    # anymore. Only supported for simple_optimizer=True so far.
+    "simple_optimizer": True,
+    "_disable_distributed_execution_api": True,
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/agents/pg/default_config.py]: done

File: rllib/agents/trainer.py (hunk 1 above)
Comment (sven1977) [reply rllib/agents/trainer.py:836]: This is already done a few lines above.

File: rllib/agents/trainer.py (hunk 2 above)
Comment (sven1977) [reply rllib/agents/trainer.py:993]: Totally! We should containerize these functionalities more.

File: rllib/agents/trainer.py (hunk 4 above)
Comment (sven1977) [reply rllib/agents/trainer.py:1031]: Yeah, `self._episodes_to_be_collected` are the ray object refs referencing those metrics results that timed out in the `ray.wait([w.get_metrics.remote() for w in self.workers])` call. Note that timeouts usually never happens as the default timeout threshold is 180sec.

Comment (sven1977) [reply rllib/agents/trainer.py:1031]: In other words, we never drop episodes. We always wait until they are completed before even sending them from the Workers to the driver and those that time out (`ray.wait()`) are collected in the next iteration.

File: rllib/evaluation/worker_set.py (hunk 8 above)
Comment (sven1977) [reply rllib/evaluation/worker_set.py]: No, you were right! :) We should do both: Put it in the object store (`ray.put`) for better efficiency, but also keep the actual weights for synching to the local_worker, e.g. of the evaluation worker set!

Comment (sven1977) [reply rllib/evaluation/worker_set.py]: fixed

File: rllib/agents/trainer.py (hunk 3 above)
Comment (sven1977) [reply rllib/agents/trainer.py]: We do. I think I did this for backward compatibility. ... :/ Yeah, the `result`structure itself needs its own cleanup PR, I think. We should talk this through though before we start making changes to this. Many users could relay on certain keys being where they expect them to be.

**Review by gjoliver — APPROVED** [review]
ok, let's give this a try.
this PR is getting a bit too large, hard to say if some things will work.
I think the logic next step is to add a CI test for this agent with exec_plan disabled.

File: rllib/agents/trainer.py [hunk 10]
```diff
@@ -1189,6 +1247,58 @@ def duration_fn(num_units_done):
...
+        self._counters[NUM_ENV_STEPS_SAMPLED] += num_env_steps
+        self._counters[NUM_AGENT_STEPS_SAMPLED] += num_agent_steps
+
+        # Combine all batches at once
+        train_batch = SampleBatch.concat_samples(sample_batches)
+
+        # Use simple optimizer (only for multi-agent or tf-eager; all other
+        # cases should use the multi-GPU optimizer, even if only using 1 GPU).
+        # TODO: (sven) rename MultiGPUOptimizer into something more
+        #  meaningful.
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/agents/trainer.py:1292]: Yes, correct. We should add multi-agent capability for the `training_iteration` function soon.
